pkg/upgrade/migrate.go

```diff
@@ -0,0 +1,341 @@
+	}
+	podIPMap := make(map[string]*corev1.Pod)
+	for i := 0; i < len(podList.Items); i++ {
+		pod := &podList.Items[i]
+		log.WithFields(log.Fields{"pod": pod.Name, "IP": pod.Status.PodIP, "namespace": pod.Namespace}).Info("mapping in pod")
+		podIPMap[pod.Status.PodIP] = pod
+	}
+	log.Info("successfully mapped pod ips to pods!")
+
+	// Acquire a lock on the host-local cni backend
+	log.Info("acquiring lock on host-local backend...")
+	hostLocal, err := disk.New(ipAllocPath, "")
+	if err != nil {
+		return fmt.Errorf("failed to initialize host local backend: %s", err)
+	}
+	if err = hostLocal.Lock(); err != nil {
+		return fmt.Errorf("failed to acquire lock on host local backend: %s", err)
+	}
+	log.Info("successfully acquired lock on host-local backend!")
+	defer func() {
+		// Release the lock on host local backend
+		log.Info("releasing lock on host-local backend...")
+		if err = hostLocal.Unlock(); err != nil {
+			log.WithError(err).Error("failed to release lock on host local backend")
+		}
+		log.Info("successfully released lock on host-local backend!")
+	}()
+
+	// Read in all the files in the host-local directory.
+	log.Info("reading files from host-local IPAM data dir...")
+	files, err := ioutil.ReadDir(ipAllocPath)
+	if err != nil {
+		return fmt.Errorf("failed to read path %s: %s", ipAllocPath, err)
+	}
+
+	// For each file, convert it into an IP allocation and then delete the file.
+	for _, f := range files {
+		logCtxt := log.WithField("file", f.Name())
+		logCtxt.Info("processing file...")
+		fname := path.Join(ipAllocPath, f.Name())
+
+		// Delete and skip the last reserved IP.
+		if strings.Contains(f.Name(), "last") {
+			logCtxt.Info("skipping and removing last reserved ip file...")
+			if err = os.Remove(fname); err != nil {
+				return fmt.Errorf("failed to remove file %s: %s", fname, err)
+			}
+			logCtxt.Info("successfully removed last reserved ip file!")
+			continue
+		}
+
+		// Skip the lock
+		if f.Name() == "lock" {
+			logCtxt.Info("skipping the lock...")
+			continue
+		}
+
+		// The name of the file is its IP address.
+		ip, _, err := cnet.ParseCIDR(fmt.Sprintf("%s/32", f.Name()))
+		if err != nil {
+			return fmt.Errorf("failed to parse IP %s: %s", f.Name(), err)
+		}
+
+		// The contents are the container ID.
+		b, err := ioutil.ReadFile(fname)
+		if err != nil {
+			return fmt.Errorf("failed to read file %s: %s", fname, err)
+		}
+		containerID := string(b)
+
+		// See if the IP hasn't already been assigned for the handleID.
+		handleID := utils.GetHandleID("k8s-pod-network", containerID, "")
+		ips, err := c.IPAM().IPsByHandle(ctxt, handleID)
+		if err == nil {
+		} else if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			// Return error if it is NOT does not exist, otherwise continue.
+			return fmt.Errorf("failed to fetch ips for handleID %s: %s", handleID, err)
+		}
+		logCtxt.WithField("handleID", handleID).Info("file contains handle ID")
+
+		alreadyExists := false
```

Comment:
IIUC, I think a more robust way to do this check would be to try to assign the IP address and if you get an AlreadyExists error, you continue.

It means you could remove a bunch of code by getting rid of the `IPsByHandle` call above altogether, as well as this cross-correlation code to check if it exists. TBH I don't really understand what this loop is meant to check, so I could definitely be wrong.